File: regions/system/library/dart/test_helper/test/test_helper_test.dart
import 'package:test/test.dart';
import 'package:k1s0_test_helper/test_helper.dart';

void main() {
  group('JwtTestHelperのテスト', () {
    final helper = JwtTestHelper(secret: 'test-secret');

    test('管理者トークンが作成されること', () {
      final token = helper.createAdminToken();
      final parts = token.split('.');
      expect(parts.length, equals(3));
      final claims = helper.decodeClaims(token);
      expect(claims, isNotNull);
      expect(claims!.sub, equals('admin'));
      expect(claims.roles, equals(['admin']));
    });

    test('ユーザートークンが作成されること', () {
      final token = helper.createUserToken('user-123', ['user']);
      final claims = helper.decodeClaims(token);
      expect(claims, isNotNull);
      expect(claims!.sub, equals('user-123'));
      expect(claims.roles, equals(['user']));
    });

    test('テナント付きトークンが作成されること', () {
      final token = helper.createToken(TestClaims(
        sub: 'svc',
        roles: ['service'],
        tenantId: 't-1',
      ));
      final claims = helper.decodeClaims(token);
      expect(claims, isNotNull);
      expect(claims!.tenantId, equals('t-1'));
    });

    test('無効なトークンでnullが返されること', () {
      expect(helper.decodeClaims('invalid'), isNull);
    });
  });

  group('MockServerBuilder', () {
    test('通知サーバーモックが構築されること', () {
      final server = MockServerBuilder.notificationServer()
          .withHealthOk()
          .withSuccessResponse('/send', '{"id":"1","status":"sent"}')
          .build();

      final health = server.handle('GET', '/health');
      expect(health, isNotNull);
      expect(health!.status, equals(200));
      expect(health.body, contains('ok'));

      final send = server.handle('POST', '/send');
      expect(send, isNotNull);
      expect(send!.status, equals(200));

      expect(server.requestCount, equals(2));
    });

    test('未知のルートでnullが返されること', () {
      final server =
          MockServerBuilder.ratelimitServer().withHealthOk().build();
      expect(server.handle('GET', '/nonexistent'), isNull);
    });

    test('エラーレスポンスが設定できること', () {
      final server = MockServerBuilder.tenantServer()
          .withErrorResponse('/create', 500)
          .build();
      final res = server.handle('POST', '/create');
      expect(res, isNotNull);
      expect(res!.status, equals(500));
      expect(res.body, contains('error'));
    });
  });

  group('FixtureBuilder', () {
    test('有効なUUIDが生成されること', () {
      final id = FixtureBuilder.uuid();
      expect(id.length, equals(36));
      expect(id, contains('-'));
    });

    test('メールアドレスが生成されること', () {
      final email = FixtureBuilder.email();
      expect(email, contains('@example.com'));
    });

    test('プレフィックス付きの名前が生成されること', () {
      final name = FixtureBuilder.name();
      expect(name, startsWith('user-'));
    });

    test('指定した範囲内の整数が生成されること', () {
      for (var i = 0; i < 100; i++) {
        final val = FixtureBuilder.intValue(min: 10, max: 20);
        expect(val, greaterThanOrEqualTo(10));
        expect(val, lessThan(20));
      }
    });

    test('最小値と最大値が同じ場合に最小値が返されること', () {
      expect(FixtureBuilder.intValue(min: 5, max: 5), equals(5));
    });

    test('テナントIDが生成されること', () {
      expect(FixtureBuilder.tenantId(), startsWith('tenant-'));
    });

    test('一意な値が生成されること', () {
      final a = FixtureBuilder.uuid();
      final b = FixtureBuilder.uuid();
      expect(a, isNot(equals(b)));
    });
  });

  group('AssertionHelper', () {
    test('JSONの部分一致で検証が通ること', () {
      AssertionHelper.assertJsonContains(
        {'id': '1', 'status': 'ok', 'extra': 'ignored'},
        {'id': '1', 'status': 'ok'},
      );
    });

    test('ネストしたJSONの部分一致で検証が通ること', () {
      AssertionHelper.assertJsonContains(
        {
          'user': {'id': '1', 'name': 'test'},
          'status': 'ok'
        },
        {
          'user': {'id': '1'}
        },
      );
    });

    test('JSONが一致しない場合に検証が失敗すること', () {
      expect(
        () => AssertionHelper.assertJsonContains({'id': '1'}, {'id': '2'}),
        throwsA(isA<AssertionError>()),
      );
    });

    test('発行されたイベントが見つかること', () {
      final events = [
        {'type': 'created', 'id': '1'},
        {'type': 'updated', 'id': '2'},
      ];
      AssertionHelper.assertEventEmitted(events, 'created');
      AssertionHelper.assertEventEmitted(events, 'updated');
    });

    test('イベントが存在しない場合に例外がスローされること', () {
      expect(
        () => AssertionHelper.assertEventEmitted(
          [{'type': 'created'}],
          'deleted',
        ),
        throwsA(isA<AssertionError>()),
      );
    });
  });
}

File: regions/system/library/typescript/test-helper/__tests__/test_helper.test.ts
import { describe, it, expect } from 'vitest';
import {
  JwtTestHelper,
  MockServerBuilder,
  FixtureBuilder,
  AssertionHelper,
} from '../src/index.js';

describe('JwtTestHelper', () => {
  const helper = new JwtTestHelper('test-secret');

  it('should create admin token', () => {
    const token = helper.createAdminToken();
    const parts = token.split('.');
    expect(parts).toHaveLength(3);
    const claims = helper.decodeClaims(token);
    expect(claims?.sub).toBe('admin');
    expect(claims?.roles).toEqual(['admin']);
  });

  it('should create user token', () => {
    const token = helper.createUserToken('user-123', ['user']);
    const claims = helper.decodeClaims(token);
    expect(claims?.sub).toBe('user-123');
    expect(claims?.roles).toEqual(['user']);
  });

  it('should create token with tenant', () => {
    const token = helper.createToken({
      sub: 'svc',
      roles: ['service'],
      tenantId: 't-1',
    });
    const claims = helper.decodeClaims(token);
    expect(claims?.tenantId).toBe('t-1');
  });

  it('should return null for invalid token', () => {
    expect(helper.decodeClaims('invalid')).toBeNull();
  });
});

describe('MockServerBuilder', () => {
  it('should build notification server mock', () => {
    const server = MockServerBuilder.notificationServer()
      .withHealthOk()
      .withSuccessResponse('/send', '{"id":"1","status":"sent"}')
      .build();

    const health = server.handle('GET', '/health');
    expect(health?.status).toBe(200);
    expect(health?.body).toContain('ok');

    const send = server.handle('POST', '/send');
    expect(send?.status).toBe(200);

    expect(server.requestCount()).toBe(2);
  });

  it('should return null for unknown route', () => {
    const server = MockServerBuilder.ratelimitServer().withHealthOk().build();
    expect(server.handle('GET', '/nonexistent')).toBeNull();
  });

  it('should support error responses', () => {
    const server = MockServerBuilder.tenantServer()
      .withErrorResponse('/create', 500)
      .build();
    const res = server.handle('POST', '/create');
    expect(res?.status).toBe(500);
    expect(res?.body).toContain('error');
  });
});

describe('FixtureBuilder', () => {
  it('should generate valid UUID', () => {
    const id = FixtureBuilder.uuid();
    expect(id).toHaveLength(36);
    expect(id).toContain('-');
  });

  it('should generate email', () => {
    const email = FixtureBuilder.email();
    expect(email).toContain('@example.com');
  });

  it('should generate name with prefix', () => {
    const name = FixtureBuilder.name();
    expect(name).toMatch(/^user-/);
  });

  it('should generate int in range', () => {
    for (let i = 0; i < 100; i++) {
      const val = FixtureBuilder.int(10, 20);
      expect(val).toBeGreaterThanOrEqual(10);
      expect(val).toBeLessThan(20);
    }
  });

  it('should return min when min equals max', () => {
    expect(FixtureBuilder.int(5, 5)).toBe(5);
  });

  it('should generate tenant id', () => {
    expect(FixtureBuilder.tenantId()).toMatch(/^tenant-/);
  });

  it('should generate unique values', () => {
    const a = FixtureBuilder.uuid();
    const b = FixtureBuilder.uuid();
    expect(a).not.toBe(b);
  });
});

describe('AssertionHelper', () => {
  it('should pass on JSON partial match', () => {
    AssertionHelper.assertJsonContains(
      { id: '1', status: 'ok', extra: 'ignored' },
      { id: '1', status: 'ok' }
    );
  });

  it('should pass on nested JSON partial match', () => {
    AssertionHelper.assertJsonContains(
      { user: { id: '1', name: 'test' }, status: 'ok' },
      { user: { id: '1' } }
    );
  });

  it('should fail on JSON mismatch', () => {
    expect(() => {
      AssertionHelper.assertJsonContains({ id: '1' }, { id: '2' });
    }).toThrow('JSON partial match failed');
  });

  it('should find emitted event', () => {
    const events = [
      { type: 'created', id: '1' },
      { type: 'updated', id: '2' },
    ];
    AssertionHelper.assertEventEmitted(events, 'created');
    AssertionHelper.assertEventEmitted(events, 'updated');
  });

  it('should throw for missing event', () => {
    expect(() => {
      AssertionHelper.assertEventEmitted([{ type: 'created' }], 'deleted');
    }).toThrow('not found');
  });
});
